### ISSUE TITLE
Convert frontend to Single-Page Application

### ISSUE THREAD

**Issue by tienne-B** [TabbycatDebate/tabbycat#1979]
Seems to be the trend now for UX and coupling concerns, the frontend and backend could be split. The frontend would make calls to the backend through the API rather than being rendered as a page within Django. This would entail a big change similar to #1380.

**Comment by czlee** [comment]
Yeah, my unqualified thoughts are as follows: I've been thinking this might be an appropriate shift for some time, since I think it's important for tab software to keep up with current software practice. (This might seem uncontroversial now, but when I entered the tabbing scene back in 2012, conventional wisdom was to change as little as possible from tried and true systems.) But as you say, it'd be a huge change. Personally, I lack the expertise to be able to make any meaningful start in it (and wasn't willing to learn it just for this). For example, I wondered, among other things, whether this would render the majority of our existing Django-based framework obsolete—and if it does, whether the more appropriate trajectory for tab software was to start over with a new project without legacy considerations.

**Comment by tienne-B** [comment]
Yes, I'd say this would make most of Django somewhat obsolete, like with all the work placed into templates/forms, and would then have to be migrated over to the frontend. However, for me the main draw of Django is the ORM and features that would still be useful when unlinked. I would suggest that it would be good to then move all the API stuff into the app to which they correspond.

It'd be some dogfooding to make sure the API works well!

**Comment by philipbelesky** [comment]
I’m sort of in two minds as to whether this a worthwhile endeavour, even without time constraints. It’s definitely a current best practice when making apps from scratch, but a lot of the driving forces there don’t apply to us - e.g. CDN/edge deployments, separation of concerns to better enable mobile apps, live data refreshes, etc. If anything we might want to consider going in a different direction where we simplify the app architecture so it is easier to maintain and deploy. 

As noted, all of our Django logic would hold, we’re really just swapping out one templating system for another and pushing template logic onto the API or something like GraphQL.  All the table view pages would be particularly easy as they are just thin wrappers around Vue tables and already supplied through JSON. Forms would tricky.

A first step here would be to improve our current Django-Vue interaction; dumping JSON into the body is probably very inefficient. Instead, all table based pages could probably be a single generic Django template (that can be cached) with a separate endpoint that the client fetches and populates from.
